### PR TITLE
Improve abi decoding functions for arrays.

### PR DESCRIPTION
**Pull request by ekpyron**
Improves gas and decreases stack pressure while actually being nicer IMHO.
[noticed this when checking where the stack errors on abi decoding usually actually come from - variables changing in loops are the worst]

### CODE REVIEW

File: libsolidity/codegen/ABIFunctions.cpp [hunk 1]
```diff
@@ -1186,11 +1186,11 @@ string ABIFunctions::abiDecodingFunctionArrayAvailableLength(ArrayType const& _t
 					mstore(array, length)
 					dst := add(array, 0x20)
 				</dynamic>
-				let src := offset
-				if gt(add(src, mul(length, <stride>)), end) {
+				let srcEnd := add(offset, mul(length, <stride>))
+				if gt(srcEnd, end) {
 					<revertInvalidStride>()
```
Comment (ekpyron) [inline libsolidity/codegen/ABIFunctions.cpp:1192]: Actually... invalid stride? Why was the revert message here ever "invalid stride"? Am I missing something here?
This is really an out-of-bounds error, isn't it?

Comment (ekpyron) [reply libsolidity/codegen/ABIFunctions.cpp:1192]: What does "invalid stride" even mean as a runtime error ;-)?

Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp:1192]: true

Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp:1192]: `length` is already checked to be non-gigantic, right?

Comment (chriseth) [reply libsolidity/codegen/ABIFunctions.cpp:1192]: Ah, indirectly through the allocation.

**Comment by ekpyron** [comment]
[pushed command line test updates]

**Review by chriseth — DISMISSED** [review]
We should really have an optimizer stage that improves this.

**Comment by ekpyron** [comment]
I just pushed some more gas cost updates - I didn't update the non-optimized run before.

**Review by chriseth — APPROVED** [review]

**Pull request merged**